### ISSUE TITLE
starcounter-include should not import empty responses

### ISSUE THREAD

**Issue by miyconst** [Starcounter/starcounter-include#12]
Right now an empty JSON object provided as a partial causes an `imported-template` with `content="http://localhost:8080/sc/htmlmerger?"`.

Server responses with `500` and `imported-template` warns about missing `<template>`.

cc @warpech, @tomalec, @chrhol

**Comment by tomalec** [comment]
The question is why do you consider it as a bu of sc-include, what does partial equal to `{}` mean?

I thought, that is what we need "`/sc/htmlmerger?` + `key=Html` for every namespace". If we need more rules, lets state them first.

If we are going to set URL=`undefined` for `{}`, what should we do for:
1. `{"only":"primitives","but":"no Html key"}`
2. `{"more":{"objects": "without"}, "Htmlproperty": {}}`

?

Corrently:
- `{}` - > `/sc/htmlmerger?`
- `{"only":"primitives","but":"no Html key"}` - > `/sc/htmlmerger?`
- `{"more":{"objects": "without"}, "Htmlproperty": {}}` - > `/sc/htmlmerger?more=&Htmlproperty=` 
  
  (to support default renderers, etc. as mentioned in https://github.com/Starcounter/Starcounter/issues/3539, https://github.com/Starcounter/starcounter-include/issues/10, https://github.com/Starcounter/starcounter-include/issues/11)

**Comment by miyconst** [comment]
On the server side the `/sc/htmlmerger?` is session less. There is no information what page this is, means there is no possibility to add an adequate response in several, but not all places.

Previously `{}` meant 'nothing in HTML', and this is widely used in SWO. For example:

``` json
"Html": "/ListPage.html",
"Items": [ { "Key": "", "Name": "" } ],
"SelectedItem": {}
```

The `SelectedItem` is a partial, which is empty by default. When such page is loaded, I don't want to see any errors on the console, and don't want to send any request to the server.

Once an item is selected, and the value is set:

``` cs
this.SelectedItem = Self.GET("/Item/" + item.Key);
```

It should be imported. Same as imported it should be removed if the `this.SelectedItem` set to `null`.

**Comment by tomalec** [comment]
> Previously {} meant 'nothing in HTML',

I think that is so, due to lack of support for `null` https://github.com/Starcounter/Starcounter/issues/3072, https://github.com/Starcounter/Starcounter/issues/3144 

## 

Could you say, what would you like to see for `{}`, `{"only":"primitives","but":"no Html key"}`, `{"more":{"objects": "without"}, "Htmlproperty": {}}`?

I would like to avoid too many heuristic, magic rules to distinguish them. Then if we would like to to set `URL` as undefined for all of them, we will fail to achieve default renderers.

**Comment by miyconst** [comment]
For all of those cases, I would like to see nothing in the HTML. If for any debugging purpose I need something else, I can add that into my template.

**Comment by tomalec** [comment]
Then we will not be able to render namespaces without `Html` property using default renderers.
And this is what we: @warpech, @chrhol, me, and Jocke wanted long time ago, and mentioned at last F2F meeting.

**Comment by miyconst** [comment]
I am not against default renderers or other features. I just try to solve daily task:

> How to define an empty partial? Previously it was possible to use `<template is="dom-if" if="{{model.Partial.Html}}">`, now it won't work as `Html` property has gone, and at the same time `{}` means `"Send an empty request to server and rise errors"`.

And a question about default renderers:

> How do we know how to render `/sc/htmlmerger?`? There is no information given, is it an empty partial, is it a partial without `Html` or a partial which does not require anything in HTML?

**Comment by tomalec** [comment]
I'm not quite sure what you mean by "empty partial". By the definition we coined last week: partial is a separated part of an application. 

If in your case, it is a part that currently is not rendering any markup, it can return HTML which is empty, like:

**/myapp/currentperson.html**

``` html
<template>
  <div class="myapp-person-card">
    <template is="dom-if" if="{{model.Person}}">
       ...
```

> `{}` means "Send an empty request to server and rise errors".

Definitely, I do agree,  this should not rise errors

> How do we know how to render /sc/htmlmerger?

It's totally up to **a** Launcher. In our case I would propose:

`/sc/htmlmerger?` - `204 No Content`
and `imported-template` should not rise an error in case of such empty response.

But if `{}` really means "content is still to come", we could also use this to implement pre-loaders, which David asked for.
`/sc/htmlmerger?` - `200 text/html <template><juicy-preloader>Content will appear soon</juicy-preloader></template>`

**Comment by tomalec** [comment]
I have added a workaround to make Launcher work with latest changes. But it's another quickfix, and I believe we are hiding a more serious bug with it. 

In general it works:
1. If there is no `partial` attribute (`partial === null`) do nothing/clear existing stamped nodes
2. If there is any `partial`
   1. Use `partial.Html` if given,
   2. Otherwise build new URL as follows 
      
      ``` js
      mergedHtmlPrefix + partial.forEach( (key, value) => key + '=' + value || defaultHtml ).join('&')
      ```

---

Then the hack is to treat `partial = {}` as `== null`.

---

@chrhol, @warpech am I correct that `{}` is there even if it should be `null`?

**Comment by miyconst** [comment]
> am I correct that `{}` is there even if it should be `null`?

Yes, that't current behavior.

**Comment by warpech** [comment]
In my opinion, if the length of the key/value array is 0, then there should be no request.

But, it there is a request, there should be no error.

**Comment by miyconst** [comment]
> In my opinion, if the length of the key/value array is 0, then there should be no request.

That was my initial idea.

> But, it there is a request, there should be no error.

I have fixed the error in this commit: https://github.com/StarcounterPrefabs/Launcher/commit/4d0724cbcd0113f46a684aab7f766bd246335b69. Now Launcher returns `<template></template>` for `/sc/htmlmerger?` request. Returning an empty string causes `<imported-template>` to swear and ask to don't misbehave by providing no template in a partial.

**Comment by tomalec** [comment]
Yesterday with @warpech we decided that
- **`partial === null`**
  
  or `Object.keys({}).length === 0` - as currently due to [server-side issues (Starcounter#3072)](https://github.com/Starcounter/Starcounter/issues/3072) we use `{}` in stead of `null`
  
  No URL will be build, no request will be sent.
- **!!partial.Html**
  
  We will fetch partial from URL given.
- **!partial.Html**
  
  We will build merged URL from **all** namespaces in `partial`.
  So `{key1: {Html: "value"}, key2: "smth w/o .Html"}`, will result in `_mergerprefix_key1=value&key2=`
  - server-side merger will be able to decide what to do with such URL (serve something for `key2` or not)
  - `imported-template` will not throw an error for empty documents

---

Thanks to that:
- we still cover server-side problems
- if someone put non-partial object (part of a view-model not fetched via `Partial.Get`), then client-side will not hide anything, but forward it to the server. So, app dev instead of error-free empty space, will notice an request being sent to server with hist object serialized. 
  
  > with JSON editors served from Launcher, for HTML less partials, it would be even more obvious, as he/she will notice the wrong object
- no errors will be thrown in supported use-cases

---

So @miyconst, I would revert your changes with empty `<template></template>` served, and add support for empty documents to `imported-template` (`204 No Content`, merged partials with empty `<imported-template-scope>`). That should give use even more flexibility in implementing `juicy-composition` features, as composer would get and area and scope for html-less app, so he/she could implement his/hers own custom, shadow DOM placeholder for it.

**Comment by miyconst** [comment]
> So @miyconst, I would revert your changes with empty <template></template> served, and add support for empty documents to imported-template ...

Please do so.

**Comment by tomalec** [comment]
I have added support to `juicy-html` and `imported-template`.

However, it seems we cannot support empty responses in end solution, as Web Components polyfill does not support importing empty resource (`onload` callback is not triggered).

We may try to cheat with adding white spaces, but I feel it may make us more harm than good.

**Comment by miyconst** [comment]
Then do you think keeping the empty template response `<template></template>` seems to be the way to go, at least for now?

**Comment by tomalec** [comment]
I have submited a PR wwith fix, could we give some time for them to accept
it?
23 maj 2016 10:40 "Konstantin" notifications@github.com napisał(a):

**Comment by miyconst** [comment]
Of course.

**Comment by warpech** [comment]
I think in the meantime it is perfectly fine to have a non empty response from HTML Merger (Starcounter and Launcher). It can be something more meaningful than whitespace - a comment:

```
<!-- Starcounter HTML merger -->
```

And in Launcher:

```
<!-- Launcher HTML merger -->
```

**Comment by tomalec** [comment]
@warpech do you mean, you would like to support importing document that have html content but doesn't have any `<template>` to import?

**Comment by warpech** [comment]
Do you mean importing the HTML comment into the master document? Right... Well, I think that we already support it, but with a warning?

**Comment by tomalec** [comment]
We do not stamp the content of imported document!

We only fetch the content of template inside. 
That's the reason to use HTML Imports, to control the dependencies and scripts.

The document given in attribute is imported (code there is executed according to HTML Imports spec) then we find the template inside, and clone it into main document.

BTW, I thought about writing article to webcomponents.org on ~"Doing partials the right way with Web Components"

**Comment by miyconst** [comment]
> BTW, I thought about writing article to webcomponents.org on ~"Doing partials the right way with Web Components"

Remember to share with us, if you do so :)

**Comment by tomalec** [comment]
Blocked on webcomponents/webcomponentsjs#535

**Comment by tomalec** [comment]
I made another PR to new repo of html-imports polyfill, which was merged, so we will see how it behaves in V1 https://github.com/webcomponents/html-imports/pull/4

**Comment by miyconst** [comment]
Won't be fixed, since the component is not relevant for Starcounter 3.
